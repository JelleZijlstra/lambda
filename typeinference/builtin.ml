open Ast

let echo_f v =
    let str = Ast.string_of_value v in
    Printf.printf "%s" str;
    VUnit

let print_f v =
    let str = Ast.string_of_value v in
    Printf.printf "%s\n" str;
    VUnit

let add_strings_f s1 =
    Ast.VBuiltin (fun s2 -> match s1, s2 with
        | VString s1', VString s2' -> VString(s1' ^ s2')
        | _ -> failwith("Typecheck failure"))

let get_builtins () =
    let m = VarMap.empty in
    (* echo : forall a. a -> () *)
    let m = VarMap.add "echo" (TForAll(["a"], TFunction(TNamedType "a", TUnit)), echo_f) m in
    (* print : forall a. a -> () *)
    let m = VarMap.add "print" (TForAll(["a"], TFunction(TNamedType "a", TUnit)), print_f) m in
    (* add_strings : string -> string -> string *)
    let m = VarMap.add "add_strings" (TFunction(TString, TFunction(TString, TString)), add_strings_f) m in
    m

let builtins = get_builtins()

let builtin_types = VarMap.map (fun (t, _) -> t) builtins

let builtin_values = VarMap.map (fun (_, v) -> v) builtins
