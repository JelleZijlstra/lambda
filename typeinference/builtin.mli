val builtins : (Ast.ltype * Ast.builtin_function) Ast.VarMap.t

val builtin_types : Ast.ltype Ast.VarMap.t

val builtin_values : Ast.builtin_function Ast.VarMap.t
