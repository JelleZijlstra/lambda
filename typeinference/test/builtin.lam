let x = "hello"

;;

echo x
