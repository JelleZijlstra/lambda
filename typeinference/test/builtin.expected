%%stdout%%
"hello"Result: module
	let x = "hello"
	let __result__ = ()
end
%%stderr%%
